File: crates/spur-notebook/src/html_video/library.rs
use std::{
    cmp::Ordering,
    error::Error,
    fmt, fs, io,
    path::{Path, PathBuf},
};

use directories::BaseDirs;
use serde::{Deserialize, Serialize};

const ENV_VAR: &str = "SPUR_HTML_VIDEO_LIBRARY";
const LIB_DIR: &str = "html-video-library";
const INDEX_FILE: &str = "index.json";
const DEFAULT_HTML_FILE_NAMES: [&str; 4] =
    ["template.html", "frame.html", "index.html", "index.htm"];
const DEFAULT_SKILL_FILE_NAMES: [&str; 2] = ["SKILL.md", "skill.md"];

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TemplateSearchResult {
    pub id: String,
    pub title: String,
    #[serde(default)]
    pub intent: String,
    #[serde(default)]
    pub summary: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    pub score: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TemplateMetadata {
    pub id: String,
    pub title: String,
    #[serde(default)]
    pub intent: String,
    #[serde(default)]
    pub summary: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Template {
    pub metadata: TemplateMetadata,
    pub html: String,
    pub skill_md: String,
}

#[derive(Debug)]
pub enum LibraryError {
    RootNotFound,
    NotFound { id: String },
    InvalidIndex { path: PathBuf, reason: String },
    Io(io::Error),
    Json(serde_json::Error),
}

impl fmt::Display for LibraryError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::RootNotFound => write!(f, "html video library root not found"),
            Self::NotFound { id } => write!(f, "html video template not found: {id}"),
            Self::InvalidIndex { path, reason } => {
                write!(f, "invalid index in {}: {reason}", path.display())
            }
            Self::Io(error) => write!(f, "html video library I/O error: {error}"),
            Self::Json(error) => write!(f, "html video library JSON error: {error}"),
        }
    }
}

impl Error for LibraryError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            Self::Io(error) => Some(error),
            Self::Json(error) => Some(error),
            Self::InvalidIndex { .. } | Self::RootNotFound | Self::NotFound { .. } => None,
        }
    }
}

impl From<io::Error> for LibraryError {
    fn from(error: io::Error) -> Self {
        Self::Io(error)
    }
}

impl From<serde_json::Error> for LibraryError {
    fn from(error: serde_json::Error) -> Self {
        Self::Json(error)
    }
}

#[derive(Debug, Deserialize)]
struct LibraryIndex {
    #[serde(default)]
    templates: Vec<IndexTemplate>,
    #[serde(default)]
    items: Vec<IndexTemplate>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct IndexTemplate {
    pub id: String,
    pub title: String,
    #[serde(default)]
    pub intent: String,
    #[serde(default)]
    pub summary: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
}

pub fn resolve_root(resource_dir: Option<&Path>) -> Option<PathBuf> {
    let home_dir = user_home_dir();
    let repo_assets = Path::new(env!("CARGO_MANIFEST_DIR")).join("assets");
    resolve_root_with(
        resource_dir,
        home_dir.as_deref(),
        Some(repo_assets.as_path()),
    )
}

fn resolve_root_with(
    resource_dir: Option<&Path>,
    home_dir: Option<&Path>,
    repo_assets_dir: Option<&Path>,
) -> Option<PathBuf> {
    if let Some(base) = std::env::var_os(ENV_VAR) {
        let candidate = PathBuf::from(&base);
        let legacy_candidate = candidate.join(LIB_DIR);
        if is_library_root(&candidate) {
            return Some(candidate);
        }
        if is_library_root(&legacy_candidate) {
            return Some(legacy_candidate);
        }
    }

    if let Some(home_dir) = home_dir {
        let candidate = home_dir.join(".spur").join(LIB_DIR);
        if is_library_root(&candidate) {
            return Some(candidate);
        }
    }

    if let Some(resource_dir) = resource_dir {
        let candidate = resource_dir.join(LIB_DIR);
        if is_library_root(&candidate) {
            return Some(candidate);
        }
    }

    if let Some(repo_assets_dir) = repo_assets_dir {
        let candidate = repo_assets_dir.join(LIB_DIR);
        if is_library_root(&candidate) {
            return Some(candidate);
        }
    }

    None
}

fn is_library_root(path: &Path) -> bool {
    path.join(INDEX_FILE).is_file()
}

fn user_home_dir() -> Option<PathBuf> {
    BaseDirs::new()
        .map(|dirs| dirs.home_dir().to_path_buf())
        .or_else(|| std::env::var_os("HOME").map(PathBuf::from))
}

pub fn search(
    intent: &str,
    top: usize,
    resource_dir: Option<&Path>,
) -> Result<Vec<TemplateSearchResult>, LibraryError> {
    let root = resolve_root(resource_dir).ok_or(LibraryError::RootNotFound)?;
    let items = load_index(&root)?;
    let tokens = tokenize(intent);
    let query_is_empty = tokens.is_empty();
    let mut results: Vec<_> = items
        .into_iter()
        .filter_map(|item| {
            let score = score(item.intent.as_str(), &tokens, &item);
            if !query_is_empty && score == 0.0 {
                None
            } else {
                Some(TemplateSearchResult {
                    id: item.id,
                    title: item.title,
                    intent: item.intent,
                    summary: item.summary,
                    tags: item.tags,
                    score,
                })
            }
        })
        .collect();

    results.sort_by(|left, right| {
        right
            .score
            .partial_cmp(&left.score)
            .unwrap_or(Ordering::Equal)
            .then_with(|| left.id.cmp(&right.id))
            .then_with(|| left.title.cmp(&right.title))
    });
    if top > 0 {
        results.truncate(top);
    }
    Ok(results)
}

pub fn get_template(id: &str, resource_dir: Option<&Path>) -> Result<Template, LibraryError> {
    let root = resolve_root(resource_dir).ok_or(LibraryError::RootNotFound)?;
    let items = load_index(&root)?;
    let item = items
        .into_iter()
        .find(|item| item.id == id)
        .ok_or_else(|| LibraryError::NotFound { id: id.to_string() })?;

    let template_dir = resolve_template_dir(&root, &item.id)?;
    let html = read_template_html(&template_dir)?;
    let skill_md = read_template_skill_md(&template_dir)?;

    Ok(Template {
        metadata: TemplateMetadata {
            id: item.id,
            title: item.title,
            intent: item.intent,
            summary: item.summary,
            tags: item.tags,
        },
        html,
        skill_md,
    })
}

fn resolve_template_dir(root: &Path, id: &str) -> Result<PathBuf, LibraryError> {
    let package_dir = root.join("templates").join(id);
    if package_dir.is_dir() {
        return Ok(package_dir);
    }

    let direct_dir = root.join(id);
    if direct_dir.is_dir() {
        Ok(direct_dir)
    } else {
        Err(LibraryError::NotFound { id: id.to_string() })
    }
}

fn load_index(root: &Path) -> Result<Vec<IndexTemplate>, LibraryError> {
    let index_path = root.join(INDEX_FILE);
    let raw = fs::read_to_string(&index_path).map_err(|error| {
        if error.kind() == io::ErrorKind::NotFound {
            LibraryError::InvalidIndex {
                path: index_path.clone(),
                reason: "missing index".to_string(),
            }
        } else {
            LibraryError::Io(error)
        }
    })?;

    let array_result: Result<Vec<IndexTemplate>, serde_json::Error> = serde_json::from_str(&raw);
    if let Ok(items) = array_result {
        return Ok(items);
    }

    let file: LibraryIndex = serde_json::from_str(&raw)?;
    let items = if !file.templates.is_empty() {
        file.templates
    } else if !file.items.is_empty() {
        file.items
    } else {
        Vec::new()
    };
    Ok(items)
}

fn read_template_html(template_dir: &Path) -> Result<String, LibraryError> {
    for name in DEFAULT_HTML_FILE_NAMES {
        let candidate = template_dir.join(name);
        if candidate.is_file() {
            return fs::read_to_string(candidate).map_err(Into::into);
        }
    }
    Err(LibraryError::NotFound {
        id: template_dir
            .file_name()
            .and_then(|name| name.to_str())
            .unwrap_or_default()
            .to_string(),
    })
}

fn read_template_skill_md(template_dir: &Path) -> Result<String, LibraryError> {
    for name in DEFAULT_SKILL_FILE_NAMES {
        let candidate = template_dir.join(name);
        if candidate.is_file() {
            return fs::read_to_string(candidate).map_err(Into::into);
        }
    }
    Err(LibraryError::NotFound {
        id: template_dir
            .file_name()
            .and_then(|name| name.to_str())
            .unwrap_or_default()
            .to_string(),
    })
}

pub fn default_render_duration(frame_count: usize, fps: u32, duration: Option<f64>) -> f64 {
    let fps = u32::max(fps, 1);
    duration
        .filter(|value| value.is_finite() && *value > 0.0)
        .unwrap_or_else(|| {
            if frame_count <= 1 {
                1.0
            } else {
                frame_count as f64 / fps as f64
            }
        })
}

fn tokenize(input: &str) -> Vec<String> {
    input
        .split_whitespace()
        .map(|token| {
            token
                .trim_matches(|c: char| !c.is_alphanumeric())
                .to_lowercase()
        })
        .filter(|token| !token.is_empty())
        .collect()
}

fn score(intent: &str, tokens: &[String], item: &IndexTemplate) -> f64 {
    if tokens.is_empty() {
        return 1.0;
    }

    let mut score = 0.0;
    let title = item.title.to_lowercase();
    let intent_index = intent.to_lowercase();
    let summary = item.summary.as_deref().unwrap_or_default().to_lowercase();
    let tags = item.tags.join(" ").to_lowercase();

    for token in tokens {
        if item.id.to_lowercase().contains(token) {
            score += 10.0;
        }
        if title.contains(token) {
            score += 6.0;
        }
        if intent_index.contains(token) {
            score += 4.0;
        }
        if !summary.is_empty() && summary.contains(token) {
            score += 1.5;
        }
        if tags.contains(token) {
            score += 0.5;
        }
    }

    score
}

File: app_gallery/html_video/server/library.py
from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any


INDEX_FILE = "index.json"
DEFAULT_HTML_FILE_NAMES = ("template.html", "frame.html", "index.html", "index.htm")
DEFAULT_SKILL_FILE_NAMES = ("SKILL.md", "skill.md")


class LibraryError(Exception):
    pass


class RootNotFound(LibraryError):
    def __str__(self) -> str:
        return "html video library root not found"


class TemplateNotFound(LibraryError):
    def __init__(self, template_id: str) -> None:
        super().__init__(template_id)
        self.template_id = template_id

    def __str__(self) -> str:
        return f"html video template not found: {self.template_id}"


class InvalidIndex(LibraryError):
    def __init__(self, path: Path, reason: str) -> None:
        super().__init__(path, reason)
        self.path = path
        self.reason = reason

    def __str__(self) -> str:
        return f"invalid index in {self.path}: {self.reason}"


class LibraryJsonError(LibraryError):
    def __init__(self, error: json.JSONDecodeError) -> None:
        super().__init__(error)
        self.error = error

    def __str__(self) -> str:
        return f"html video library JSON error: {self.error}"


@dataclass(frozen=True)
class IndexTemplate:
    id: str
    title: str
    intent: str = ""
    summary: str | None = None
    tags: tuple[str, ...] = ()


def resolve_root() -> Path | None:
    raw = os.environ.get("TEMPLATES_DIR")
    if raw:
        candidate = Path(raw)
        if _is_library_root(candidate):
            return candidate

    server_dir = Path(__file__).resolve().parent
    for candidate in (server_dir / "templates", server_dir.parent / "templates"):
        if _is_library_root(candidate):
            return candidate

    return None


def search(intent: str, top: int) -> list[dict[str, Any]]:
    root = resolve_root()
    if root is None:
        raise RootNotFound()

    tokens = tokenize(intent)
    query_is_empty = len(tokens) == 0
    results: list[dict[str, Any]] = []
    for item in _load_index(root):
        item_score = score(item.intent, tokens, item)
        if not query_is_empty and item_score == 0.0:
            continue
        results.append(
            {
                "id": item.id,
                "title": item.title,
                "intent": item.intent,
                "summary": item.summary,
                "tags": list(item.tags),
                "score": item_score,
            }
        )

    results.sort(key=lambda item: (-item["score"], item["id"], item["title"]))
    if top > 0:
        del results[top:]
    return results


def get_template(template_id: str) -> dict[str, Any]:
    root = resolve_root()
    if root is None:
        raise RootNotFound()

    items = _load_index(root)
    item = next((candidate for candidate in items if candidate.id == template_id), None)
    if item is None:
        raise TemplateNotFound(template_id)

    template_dir = _resolve_template_dir(root, item.id)
    metadata = {
        "id": item.id,
        "title": item.title,
        "intent": item.intent,
        "summary": item.summary,
        "tags": list(item.tags),
    }
    return {
        "metadata": metadata,
        "html": _read_template_html(template_dir),
        "skill_md": _read_template_skill_md(template_dir),
    }


def tokenize(value: str) -> list[str]:
    return [
        token
        for token in (
            re.sub(r"^\W+|\W+$", "", chunk, flags=re.UNICODE).lower()
            for chunk in value.split()
        )
        if token
    ]


def score(intent: str, tokens: list[str], item: IndexTemplate) -> float:
    if not tokens:
        return 1.0

    total = 0.0
    item_id = item.id.lower()
    title = item.title.lower()
    intent_index = intent.lower()
    summary = (item.summary or "").lower()
    tags = " ".join(item.tags).lower()

    for token in tokens:
        if token in item_id:
            total += 10.0
        if token in title:
            total += 6.0
        if token in intent_index:
            total += 4.0
        if summary and token in summary:
            total += 1.5
        if token in tags:
            total += 0.5

    return total


def _is_library_root(path: Path) -> bool:
    return (path / INDEX_FILE).is_file()


def _load_index(root: Path) -> list[IndexTemplate]:
    index_path = root / INDEX_FILE
    try:
        raw = index_path.read_text()
    except FileNotFoundError as error:
        raise InvalidIndex(index_path, "missing index") from error

    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as error:
        raise LibraryJsonError(error) from error

    if isinstance(parsed, list):
        raw_items = parsed
    elif isinstance(parsed, dict):
        templates = parsed.get("templates") or []
        items = parsed.get("items") or []
        raw_items = templates if templates else items
    else:
        raise InvalidIndex(index_path, "expected array or object")

    return [_parse_index_template(item, index_path) for item in raw_items]


def _parse_index_template(raw: Any, index_path: Path) -> IndexTemplate:
    if not isinstance(raw, dict):
        raise InvalidIndex(index_path, "template entry must be an object")

    try:
        template_id = raw["id"]
        title = raw["title"]
    except KeyError as error:
        raise InvalidIndex(index_path, f"missing template field: {error.args[0]}") from error

    if not isinstance(template_id, str) or not isinstance(title, str):
        raise InvalidIndex(index_path, "template id and title must be strings")

    tags = raw.get("tags", [])
    if tags is None:
        tags = []
    if not isinstance(tags, list):
        raise InvalidIndex(index_path, "template tags must be an array")

    summary = raw.get("summary")
    if summary is not None and not isinstance(summary, str):
        raise InvalidIndex(index_path, "template summary must be a string")

    intent = raw.get("intent", "")
    if not isinstance(intent, str):
        raise InvalidIndex(index_path, "template intent must be a string")

    return IndexTemplate(
        id=template_id,
        title=title,
        intent=intent,
        summary=summary,
        tags=tuple(str(tag) for tag in tags),
    )


def _resolve_template_dir(root: Path, template_id: str) -> Path:
    package_dir = root / "templates" / template_id
    if package_dir.is_dir():
        return package_dir

    direct_dir = root / template_id
    if direct_dir.is_dir():
        return direct_dir

    raise TemplateNotFound(template_id)


def _read_template_html(template_dir: Path) -> str:
    for name in DEFAULT_HTML_FILE_NAMES:
        candidate = template_dir / name
        if candidate.is_file():
            return candidate.read_text()
    raise TemplateNotFound(template_dir.name)


def _read_template_skill_md(template_dir: Path) -> str:
    for name in DEFAULT_SKILL_FILE_NAMES:
        candidate = template_dir / name
        if candidate.is_file():
            return candidate.read_text()
    raise TemplateNotFound(template_dir.name)
